src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs

```diff
@@ -842,67 +795,88 @@ private static Imports BuildImports(CSharpCompilation compilation, ImmutableArra
             var usingAliases = new Dictionary<string, AliasAndUsingDirective>();
             var usingsBuilder = ArrayBuilder<NamespaceOrTypeAndUsingDirective>.GetInstance();
 
-            foreach (var importString in importStrings)
+            foreach (var importRecord in importRecords)
             {
-                string alias;
-                string externAlias;
-                string target;
-                ImportTargetKind kind;
-                if (!CustomDebugInfoReader.TryParseCSharpImportString(importString, out alias, out externAlias, out target, out kind))
-                {
-                    Debug.WriteLine("Unable to parse import string '{0}'", (object)importString);
-                    continue;
-                }
-
-                switch (kind)
+                switch (importRecord.TargetKind)
                 {
                     case ImportTargetKind.Type:
-                        {
-                            Debug.Assert(target != null, string.Format("Type import string '{0}' has no target", importString));
-                            Debug.Assert(externAlias == null, string.Format("Type import string '{0}' has an extern alias (should be folded into target)", importString));
+                    {
+                        var portableImportRecord = importRecord as PortableImportRecord;
 
-                            TypeSymbol typeSymbol = metadataDecoder.GetTypeSymbolForSerializedType(target);
-                            Debug.Assert((object)typeSymbol != null);
-                            if (typeSymbol.IsErrorType())
-                            {
-                                // Type is unrecognized. The import may have been
-                                // valid in the original source but unnecessary.
-                                continue; // Don't add anything for this import.
-                            }
-                            else if (alias == null && !typeSymbol.IsStatic)
-                            {
-                                // Only static types can be directly imported.
-                                continue;
-                            }
+                        TypeSymbol typeSymbol = portableImportRecord != null
+                                ? portableImportRecord.GetTargetType(metadataDecoder)
+                                : metadataDecoder.GetTypeSymbolForSerializedType(importRecord.TargetString);
+                        Debug.Assert((object)typeSymbol != null);
```

Comment:
Could method GetTargetType(metadataDecoder) be abstract on ImportRecord, so that we can avoid special-casing here?